db/transaction_log_impl.h

```diff
@@ -54,14 +56,80 @@ class WalFileImpl : public WalFile {
   uint64_t sizeFileBytes_;
 };
 
+class TransactionLogSeqCache {
+ public:
+  TransactionLogSeqCache(size_t size, SystemClock* clock)
+      : size_(size), clock_(clock) {}
+  struct SeqWithFileBlockIdx {
+    uint64_t log_number;
+    uint64_t seq_number;
+    uint64_t block_index;
+    uint64_t timestamp;
+    SeqWithFileBlockIdx(uint64_t log_num, uint64_t seq_num, uint64_t block_idx,
+                        uint64_t ts)
+        : log_number(log_num),
+          seq_number(seq_num),
+          block_index(block_idx),
+          timestamp(ts){};
+    bool operator<(const SeqWithFileBlockIdx& other) const {
+      return std::tie(other.log_number, other.seq_number) <
+             std::tie(log_number, seq_number);
+    }
+  };
+
+  void Insert(uint64_t log_number, uint64_t seq_number, uint64_t block_index) {
+    std::lock_guard<std::mutex> lk{mutex_};
+    auto now = clock_->NowMicros();
+    auto res = cache_.emplace(log_number, seq_number, block_index, now);
+    if (!res.second) {
+      // block idx should be the same with the same log number and seq number
+      assert(block_index == res.first->block_index);
+      auto pos_hint = res.first;
+      pos_hint++;
+      cache_.erase(res.first);
+      cache_.emplace_hint(pos_hint, log_number, seq_number, block_index, now);
+    }
+    // delete the oldest record when cache is full
+    if (cache_.size() > size_) {
+      auto iter = std::min_element(cache_.begin(), cache_.end(),
+                                   [](const auto& a, const auto& b) {
+                                     return a.timestamp < b.timestamp;
+                                   });
+      cache_.erase(iter);
+    }
+  }
+
+  bool Lookup(uint64_t log_number, uint64_t seq_number, uint64_t* block_index) {
```

Comment:
This API has a returned bool that is not used, maybe consolidate its contract, or check that return value at the call sites.